test_collections/sdk_tests/support/python_testing/models/test_case.py

```diff
@@ -0,0 +1,281 @@
+    test_engine_logger,
+)
+from app.test_engine.models import TestCase, TestStep
+
+from .python_test_models import PythonTest
+from .python_testing_hooks_proxy import SDKPythonTestRunnerHooks
+
+# Custom type variable used to annotate the factory method in PythonTestCase.
+T = TypeVar("T", bound="PythonTestCase")
+
+# Command line params
+RUNNER_CLASS = "external_runner.py"
+RUNNER_CLASS_PATH = "./testing2/"
+EXECUTABLE = "python3"
+
+
+class PythonTestCase(TestCase):
+    """Base class for all Python Test based test cases.
+
+    This class provides a class factory that will dynamically declare a new sub-class
+    based on the test-type the Python test is expressing.
+
+    The PythonTest will be stored as a class property that will be used at run-time
+    in all instances of such subclass.
+    """
+
+    python_test: PythonTest
+    python_test_version: str
+    test_finished: bool
+
+    def __init__(self, test_case_execution: TestCaseExecution) -> None:
+        super().__init__(test_case_execution=test_case_execution)
+        self.chip_tool: ChipTool
+
+    def reset(self) -> None:
+        self.start_called = False
+        self.stop_called = False
+        self.test_start_called = False
+        self.test_stop_called = False
+        self.step_success_count = 0
+        self.step_failure_count = 0
+        self.step_unknown_count = 0
+        self.__runned = 0
+
+    def start(self, count: int) -> None:
+        pass
+
+    def stop(self, duration: int) -> None:
+        pass
+
+    def test_start(self, filename: str, name: str, count: int) -> None:
+        pass
+        # Dont know if it is necessary for python testing (came from chip_tool)
+        # self.next_step()
+
+    def test_stop(self, exception: Exception, duration: int) -> None:
+        self.current_test_step.mark_as_completed()
+
+    def step_skipped(self, name: str, expression: str) -> None:
+        self.current_test_step.mark_as_not_applicable(
+            f"Test step skipped: {name}. {expression} == False"
+        )
+        self.next_step()
+
+    def step_start(self, name: str) -> None:
+        pass
+
+    def step_success(self, logger: Any, logs: str, duration: int, request: Any) -> None:
+        self.__handle_logs(logs)
+        self.next_step()
+
+    def step_failure(
+        self, logger: Any, logs: str, duration: int, request: Any, received: Any
+    ) -> None:
+        self.__handle_logs(logs)
+        self.__report_failures(logger, request, received)
+        self.next_step()
+
+    def step_unknown(self) -> None:
+        self.__runned += 1
+
+    def is_finished(self) -> bool:
+        return self.test_finished
+
+    def __handle_logs(self, logs: Any) -> None:
+        for log_entry in logs or []:
+            if not isinstance(log_entry, MatterLog):
+                continue
+
+            test_engine_logger.log(
+                CHIPTOOL_LEVEL,
+                CHIP_LOG_FORMAT.format(log_entry.module, log_entry.message),
+            )
```

Comment:
`logs` is defined as `str` in the methods that call this, not a list. How would this work?